master/internal/db/postgres_tasks.go

```diff
@@ -116,6 +120,74 @@ func (db *PgDB) CompleteTask(tID model.TaskID, endTime time.Time) error {
 	return completeTask(db.sql, tID, endTime)
 }
 
+// CompleteGenericTask persists the completion of a task of type GENERIC.
+func (db *PgDB) CompleteGenericTask(tID model.TaskID, endTime time.Time) error {
+	err := completeTask(db.sql, tID, endTime)
+	if err != nil {
+		return err
+	}
+	if _, err := db.sql.Exec(`
+UPDATE tasks
+SET task_state = (
+    CASE WHEN task_state=$2 THEN $3::task_state
+    ELSE $4::task_state END)
+WHERE task_id = $1
+    `, tID, model.TaskStateStoppingCanceled, model.TaskStateCanceled, model.TaskStateCompleted); err != nil {
+		return errors.Wrap(err, "completing task")
+	}
+	return nil
+}
+
+// KillGenericTask persists the termination of a task of type GENERIC.
+func (db *PgDB) KillGenericTask(tID model.TaskID, endTime time.Time) error {
+	err := completeTask(db.sql, tID, endTime)
+	if err != nil {
+		return err
+	}
+	if _, err := db.sql.Exec(`
+UPDATE tasks
+SET task_state = $2
+WHERE task_id = $1
+	`, tID, model.TaskStateCanceled); err != nil {
+		return errors.Wrap(err, "killing task")
+	}
+	return nil
+}
+
+func (db *PgDB) SetPausedState(taskID model.TaskID, endTime time.Time) error {
+	if _, err := db.sql.Exec(`
+	UPDATE tasks
+	SET task_state = $1, end_time = $2
+	WHERE task_id = $3`, model.TaskStatePaused, endTime, taskID); err != nil {
+		return errors.Wrap(err, "pausing task")
+	}
+	return nil
+}
+
+// IsPaused returns true if given task is in paused/pausing state.
+func (db *PgDB) IsPaused(ctx context.Context, tID model.TaskID) (bool, error) {
+	query := fmt.Sprintf(`
+SELECT count(*)
+FROM tasks
+WHERE task_id = '%s' AND (task_state='%s' OR task_state='%s')
+`, tID, model.TaskStateStoppingPaused, model.TaskStatePaused)
+	var count int
+	if err := db.sql.QueryRow(query).Scan(&count); err != nil {
+		return false, err
+	}
+	return count > 0, nil
+}
+
+func (db *PgDB) SetErrorState(taskID model.TaskID, endTime time.Time) error {
+	if _, err := db.sql.Exec(`
+	UPDATE tasks
+	SET task_state = $1, end_time = $2
+	WHERE task_id = $3`, model.TaskStateError, endTime, taskID); err != nil {
+		return errors.Wrap(err, "setting error task state")
+	}
+	return nil
+}
```

Comment:
1. please use Bun (NewRaw, or actual models whenever possible) instead of the old database access methods.
2. there's one call here which does `fmt.Sprintf` to make a query - this is an SQL injection vector, and so it's extremely dangerous and almost never a good idea to use it. Use parameter placeholders (e.g. `task_state = ?`) instead.

Reply:
1. Since these functions are used in `onAllocationExit`, `ctx` is not available for these functions. Would there be an issue then for using the old db methods?
2. I can fix the `fmt.Sptrint`. There are other examples in this file we should create a ticket to fix this.

Reply:
added `context.Background()`